File: Code/CH5 Code/C#/ContainedControls-c/SomePart.cs
using System;
using System.Collections;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Windows.Forms;

namespace ContainedControls_c
{
  /// <summary>
  /// This class knows about itself
  /// </summary>
	public class SomePart
	{
    private ContextMenu mnu = new ContextMenu();
    private Decimal     mPrice;
    private string      mIdentifier;
    private string      mToolsNeeded;
    private ArrayList   mAr;
    private Image       mImg;
    static  int         mPartsCount=0;

    private struct menu_item
    {
      public MenuItem m;
      public string   s;
      public void DoSomething(string Title)
      {
        if(s != null)
          MessageBox.Show(s, Title);
        else
          MessageBox.Show("No Action", Title);
      }
    }

		public SomePart()
		{
      mPartsCount += 1;
      GetData();
      InitMenu();
    }

    // This private function is supposed to get the record for 
    // this part from the database and set up whatever fields are necessary
    private void GetData()
    {
      //Fake getting this from the database
      mPrice = 54.952M + mPartsCount;
      mToolsNeeded = "Linemans pliers\n Phillips screwdriver\n Half a brain";
      mIdentifier = "Some Part #" + mPartsCount.ToString();

      mImg = Image.FromFile(mPartsCount.ToString() + ".ico");

    }

    private void InitMenu()
    {
      mAr = new ArrayList();
      menu_item m;
      
      m = new menu_item();
      m.m = new MenuItem("Customer Price", new EventHandler(this.MenuHandler));
      m.s = "Customer Price = " + mPrice.ToString("C");
      mAr.Add(m);

      m = new menu_item();
      m.m = new MenuItem("Tools Needed", new EventHandler(this.MenuHandler));
      m.s = "Suggested Tools = " + mToolsNeeded;
      mAr.Add(m);

      for(int k=0; k<mPartsCount; k++)
      {
        m = new menu_item();
        m.m = new MenuItem("Item #" + k.ToString(), new EventHandler(this.MenuHandler));
        mAr.Add(m);
      }

      mAr.TrimToSize();

    }

    private void MenuHandler(object sender, EventArgs e)
    {
      MenuItem m = (MenuItem)sender;

      foreach(menu_item menu in mAr)
      {
        if (menu.m == m)
          menu.DoSomething(mIdentifier);
      }
    }

    public void ShowMenu(Control c, Point p)
    {
      foreach(menu_item m in mAr)
      {
        mnu.MenuItems.Add(m.m);
      }

      mnu.Show(c, p);
      mnu.MenuItems.Clear();
    }

    public Decimal Price { get{ return mPrice; } }
    public string  ID    { get{ return mIdentifier; } }
    public Image   img   { get{ return mImg; } }

	}
}

File: Code/CH5 Code/VB/ContainedControls-vb/SomePart.vb
Option Strict On

Imports System
Imports System.Drawing
Imports System.Drawing.Drawing2D
Imports System.Collections
Imports System.Windows.Forms

Public Class SomePart

  Private mnu As ContextMenu = New ContextMenu()
  Private mPrice As Decimal
  Private mIdentifier As String
  Private mToolsNeeded As String
  Private mAr As ArrayList
  Private mImg As Image
  Shared mPartsCount As Int32 = 0

  Private Structure menu_item
    Public m As MenuItem
    Public s As String
    Public Sub DoSomething(ByVal Title As String)
      If Not s Is Nothing Then
        MessageBox.Show(s, Title)
      Else
        MessageBox.Show("No Action", Title)
      End If
    End Sub
  End Structure

  Public Sub New()
    mPartsCount += 1
    GetData()
    InitMenu()
  End Sub

  ' This private function is supposed to get the record for 
  ' this part from the database and set up whatever fields are necessary
  Private Sub GetData()
    'Fake getting this from the database
    mPrice = CType(54.952, Decimal) + mPartsCount
    mToolsNeeded = "Linemans pliers\n Phillips screwdriver\n Half a brain"
    mIdentifier = "Some Part #" + mPartsCount.ToString()

    mImg = Image.FromFile(mPartsCount.ToString() + ".ico")

  End Sub

  Private Sub InitMenu()
    mAr = New ArrayList()
    Dim m As menu_item
    Dim k As Int32

    m = New menu_item()
    m.m = New MenuItem("Customer Price", New EventHandler(AddressOf MenuHandler))
    m.s = "Customer Price = " + mPrice.ToString("C")
    mAr.Add(m)

    m = New menu_item()
    m.m = New MenuItem("Tools Needed", New EventHandler(AddressOf MenuHandler))
    m.s = "Suggested Tools = " + mToolsNeeded
    mAr.Add(m)

    For k = 0 To mPartsCount - 1
      m = New menu_item()
      m.m = New MenuItem("Item #" + k.ToString(), New EventHandler(AddressOf MenuHandler))
      mAr.Add(m)
    Next

    mAr.TrimToSize()

  End Sub

  Private Sub MenuHandler(ByVal sender As Object, ByVal e As EventArgs)
    Dim m As MenuItem = DirectCast(sender, MenuItem)
    Dim menu As menu_item

    For Each menu In mAr
      If Object.Equals(menu.m, m) Then
        menu.DoSomething(mIdentifier)
      End If
    Next
  End Sub

  Public Sub ShowMenu(ByVal c As Control, ByVal p As Point)
    Dim m As menu_item

    For Each m In mAr
      mnu.MenuItems.Add(m.m)
    Next

    mnu.Show(c, p)
    mnu.MenuItems.Clear()
  End Sub

  Public ReadOnly Property Price() As Decimal
    Get
      Return mPrice
    End Get
  End Property

  Public ReadOnly Property ID() As String
    Get
      Return mIdentifier
    End Get
  End Property

  Public ReadOnly Property img() As Image
    Get
      Return mImg
    End Get
  End Property

End Class
